File: algorithms/uva/10107.cpp
#include <queue>
#include <vector>
#include <iostream>
using namespace std;

int main(){
    long x, sol, aux;
    priority_queue<long> lowq;
    priority_queue<long, vector<long>, greater<long>> highq;

    cin >> x;
    lowq.push(x);
    cout << x << endl;
    while(cin >> x){
        if(x<=lowq.top()){
            lowq.push(x);
            if(lowq.size()>highq.size()+1){
                aux = lowq.top();
                lowq.pop();
                highq.push(aux);
            }
        }else{
            highq.push(x);
            if(highq.size()>lowq.size()+1){
                aux = highq.top();
                highq.pop();
                lowq.push(aux);
            }
        }
        
        if(lowq.size()==highq.size()){
            sol = (lowq.top()+highq.top())/2;
            cout << sol << endl;
        }else if(lowq.size()>highq.size()){
            cout << lowq.top() << endl;
        }else{
            cout << highq.top() << endl;
        }
    }
    return 0;
}

File: rodri/uva/10107.py
from queue import PriorityQueue

topq = PriorityQueue()
lowq = PriorityQueue()
inp1 = int(input())
print(inp1)
inp2 = int(input())
print(int((inp1 + inp2) / 2))
if inp1 > inp2:
    topq.put(inp1)
    lowq.put(-inp2)
else:
    topq.put(inp2)
    lowq.put(-inp1)
i = 3
while True:

    try:
        inp = int(input())
        l = -lowq.get()
        t = topq.get()

        if inp <= l:
            if i % 2 == 1:
                print(l)
                lowq.put(-l)
                lowq.put(-inp)
                topq.put(t)
            else:
                lowq.put(-inp)
                aux = -lowq.get()
                lowq.put(-aux)
                topq.put(l)
                topq.put(t)
                print(int((l + aux) / 2))
        elif inp > l and inp < t:
            if i % 2 == 1:
                print(inp)
                lowq.put(-l)
                lowq.put(-inp)
                topq.put(t)
            else:
                lowq.put(-l)
                topq.put(inp)
                topq.put(t)
                print(int((l + inp) / 2))
        else:
            if i % 2 == 1:
                print(t)
                lowq.put(-l)
                lowq.put(-t)
                topq.put(inp)
            else:
                print(int((t + l) / 2))
                lowq.put(-l)
                topq.put(t)
                topq.put(inp)
        i += 1
    except:
        break